runelite-client/src/main/java/net/runelite/client/plugins/crowdsourcing/varbits/CrowdsourcingVarbits.java

```diff
@@ -0,0 +1,193 @@
+import net.runelite.api.coords.WorldPoint;
+import net.runelite.api.events.GameStateChanged;
+import net.runelite.api.events.VarbitChanged;
+import net.runelite.client.callback.ClientThread;
+import net.runelite.client.eventbus.Subscribe;
+
+@Slf4j
+public class CrowdsourcingVarbits
+{
+
+	@Inject
+	private Client client;
+
+	@Inject
+	private CrowdsourcingManager crowdsourcingManager;
+
+	@Inject
+	private ClientThread clientThread;
+
+	private static final int VARBITS_ARCHIVE_ID = 14;
+	private static final int VARBIT = 0;
+	private static final int VARPLAYER = 1;
+
+	private int[] oldVarps = null;
+	private int[] oldVarps2 = null;
+	private Multimap<Integer, Integer> varbits;
+
+	private int initializingTick = 0;
+
+	private static HashSet<Integer> blackList;
+
+	public void startUp()
+	{
+
+		/* Blacklist certain common varbs that give us little useful data.
+		 * 357 - Equipped weapon type
+		 * 5983 - Dialogue option appear/disappear
+		 */
+		blackList = new HashSet<>();
+		blackList.add(357);
+		blackList.add(5983);
+		varbits = HashMultimap.create();
+
+		if (oldVarps == null)
+		{
+			oldVarps = new int[client.getVarps().length];
+			oldVarps2 = new int[client.getVarps().length];
+		}
+
+		// Set oldVarps to be the current varps
+		System.arraycopy(client.getVarps(), 0, oldVarps, 0, oldVarps.length);
+		System.arraycopy(client.getVarps(), 0, oldVarps2, 0, oldVarps2.length);
+
+		// For all varbits, add their ids to the multimap with the varp index as their key
+		clientThread.invoke(() ->
+		{
+			if (client.getIndexConfig() == null)
+			{
+				return false;
+			}
+			IndexDataBase indexVarbits = client.getIndexConfig();
+			final int[] varbitIds = indexVarbits.getFileIds(VARBITS_ARCHIVE_ID);
+			for (int id : varbitIds)
+			{
+				VarbitComposition varbit = client.getVarbit(id);
+				if (varbit != null)
+				{
+					varbits.put(varbit.getIndex(), id);
+				}
+			}
+			return true;
+		});
+	}
+
+	public void shutDown()
+	{
+		varbits = null;
+		oldVarps = null;
+		oldVarps2 = null;
+		blackList = null;
+	}
+
+	@Subscribe
+	public void onGameStateChanged(GameStateChanged gameStateChanged)
+	{
+		if (gameStateChanged.getGameState().equals(GameState.HOPPING)
+			|| gameStateChanged.getGameState().equals(GameState.LOGGING_IN))
+		{
+			initializingTick = client.getTickCount();
+			shutDown();
+			startUp();
+		}
+	}
+
+	private void pushVarChange(int varType, int varIndex, int oldValue, int newValue, int tick)
+	{
+		/* Wait a tick before grabbing location.
+		 *
+		 * This seems to cause fewer issues than not waiting a tick.
+		 * Only noticeable issues seem to be in places like Dorgesh-Kaan, where you can go up and down
+		 * stairs quickly to trick it into thinking varbs were updated in a different location.
+		 * Without waiting a tick, certain loads/instances are messed up, including the Dorgesh-Kaan light
+		 * varbs when loading a part of the map you haven't been to.
+		 */
+		clientThread.invokeLater(() ->
```

Comment:
This doesn't delay the invoked runnable by a game tick, it delays it by a client tick. (See `Constants.CLIENT_TICK_LENGTH` and `Constants.GAME_TICK_LENGTH`) Hence I don't believe this comment is accurate.

Reply:
The issue seems to happen only on certain plane changes. If I go to Dorgesh Kaan ground floor, then go up a floor, the varb changes recorded get the worldpoint of the ground floor instead of the previous plane. This only seems to happen the first time I load the floor though. Waiting for the next client tick seems to solve this problem. But yeah, I definitely was under the impression this was waiting a game tick rather than a client tick, so I can clean up that comment.